python/ray/rllib/agents/sac/models.py

```diff
@@ -0,0 +1,179 @@
+
+        squash_bijector = (
+            SquashBijector()
+            if self._squash
+            else tfp.bijectors.Identity())
+
+        actions = tf.keras.layers.Lambda(
+            lambda raw_actions: squash_bijector.forward(raw_actions)
+        )(raw_actions)
+
+        self.actions_model = tf.keras.Model(self.condition_inputs, actions)
+
+        def log_pis_fn(inputs):
+            shift, log_scale_diag, actions = inputs
+            base_distribution = tfp.distributions.MultivariateNormalDiag(
+                loc=tf.zeros(output_shape),
+                scale_diag=tf.ones(output_shape))
+            bijector = tfp.bijectors.Chain((
+                squash_bijector,
+                tfp.bijectors.Affine(
+                    shift=shift,
+                    scale_diag=tf.exp(log_scale_diag)),
+            ))
+            distribution = (
+                tfp.distributions.ConditionalTransformedDistribution(
+                    distribution=base_distribution,
+                    bijector=bijector))
+
+            log_pis = distribution.log_prob(actions)[:, None]
+            return log_pis
+
+        self.actions_input = tf.keras.layers.Input(shape=output_shape)
+
+        log_pis = tf.keras.layers.Lambda(
+            log_pis_fn)([shift, log_scale_diag, self.actions_input])
+
+        self.log_pis_model = tf.keras.Model(
+            (*self.condition_inputs, self.actions_input), log_pis)
+
+        # self.diagnostics_model = tf.keras.Model(
+        #     self.condition_inputs,
+        #     (shift, log_scale_diag, log_pis, raw_actions, actions))
+
+    @property
+    def trainable_variables(self):
+        return self.actions_model.trainable_variables
+
+    def reset(self):
+        pass
+
+    def actions(self, conditions):
+        return self.actions_model(conditions)
+
+    def log_pis(self, conditions, actions):
+        return self.log_pis_model([*conditions, actions])
+
+    def actions_np(self, conditions):
+        return self.actions_model.predict(conditions)
+
+    def log_pis_np(self, conditions, actions):
+        return self.log_pis_model.predict([*conditions, actions])
+
+    def get_weights(self, *args, **kwargs):
+        return self.actions_model.get_weights(*args, **kwargs)
+
+    def set_weights(self, *args, **kwargs):
+        return self.actions_model.set_weights(*args, **kwargs)
+
+    def get_diagnostics(self, iteration, batch):
+        """Return diagnostic information of the policy.
+
+        Returns the mean, min, max, and standard deviation of means and
+        covariances.
+        """
+        return OrderedDict({})
+
+
+def feedforward_model(input_shape,
+                      hidden_layer_sizes,
+                      output_size,
+                      activation='relu',
+                      output_activation='linear',
+                      name=None,
+                      *args,
+                      **kwargs):
+    input_ = tf.keras.layers.Input(shape=input_shape)
```

Comment:
I currently use a custom keras model for the value functions. Might be better to use the ray-provided value functions. The keras models are quite simple and seem to become a standard way of building things in tf 2.0, which is why I've preferred to use them in our own code. Any thoughts?

Reply:
The keras models look pretty clean. Is it possible to use those inside the rllib.models.Model interface, or are there incompatibilities? I'm open to extending the model interface to support keras models if needed.

Reply:
Sounds good. I'll see if anything needs to be changed to be able to use them inside the Model.